👍 for this change, but `None` is a supported value to `reduce` in functools and I think it should be here:

```python
>>> reduce(lambda x, y: x + y, [])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: reduce() of empty sequence with no initial value
>>> reduce(lambda x, y: x + y, [], None)
```

The neatest way I know to do this is to check the length of args:

```python
def tree_reduce(*args):
  if len(args) == 2:
    f, seq = args
    return reduce(f, tree_leaves(seq))
  else:
    f, seq, default = args
    return reduce(f, seq, default)
```

Or to test for a non-None sentinel:

```python
raise_on_empty = object()

def tree_reduce(f, seq, default=raise_on_empty):
  seq = tree_leaves(seq)
  if default is raise_on_empty:
    return reduce(f, seq)
  else:
    return reduce(f, seq, default)
```

--- a/jax/tree_util.py
+++ b/jax/tree_util.py
@@ -231,10 +231,13 @@ def _replace_nones(sentinel, tree):
     else:
       return tree
 
-def tree_reduce(f, tree, initializer=None):
-  if initializer is not None:
-    return functools.reduce(f, tree_leaves(tree), initializer)
-  return functools.reduce(f, tree_leaves(tree))
+def tree_reduce(*args):
+  if len(args) == 2:
+    f, seq = args
+    return reduce(f, tree_leaves(seq))
+  else:
+    f, seq, initializer = args
+    return reduce(f, seq, initializer)
 
 def tree_all(tree):
   return all(tree_leaves(tree))

--- a/jax/tree_util.py
+++ b/jax/tree_util.py
@@ -233,11 +233,11 @@ def _replace_nones(sentinel, tree):
 
 def tree_reduce(*args):
   if len(args) == 2:
-    f, seq = args
-    return reduce(f, tree_leaves(seq))
+    f, tree = args
+    return reduce(f, tree_leaves(tree))
   else:
-    f, seq, initializer = args
-    return reduce(f, seq, initializer)
+    f, tree, initializer = args
+    return reduce(f, tree_leaves(tree), initializer)
 
 def tree_all(tree):
   return all(tree_leaves(tree))

--- a/jax/tree_util.py
+++ b/jax/tree_util.py
@@ -234,10 +234,10 @@ def _replace_nones(sentinel, tree):
 def tree_reduce(*args):
   if len(args) == 2:
     f, tree = args
-    return reduce(f, tree_leaves(tree))
+    return functools.reduce(f, tree_leaves(tree))
   else:
     f, tree, initializer = args
-    return reduce(f, tree_leaves(tree), initializer)
+    return functools.reduce(f, tree_leaves(tree), initializer)
 
 def tree_all(tree):
   return all(tree_leaves(tree))

